README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
ICcforest
=========

The goal of ICcforest is to implement the conditional inference forest approach to modeling interval-censored survival data. It also provides functions to tune the parameters and evaluate the model fit.

Installation
------------

You can install the released version of ICcforest from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("ICcforest")
```

Example
-------

This is a basic example which shows you how to solve a common problem:

``` r
## basic example code with miceData
library(ICcforest)
library(survival)
library(icenReg)
#> Loading required package: Rcpp
#> Loading required package: coda
data(miceData)

## For ICcforest to run, Inf should be set to be a large number, for example, 9999999.
idx_inf <- (miceData$u == Inf)
miceData$u[idx_inf] <- 9999999.

## Fit an iterval-censored conditional inference forest
Cforest <- ICcforest(Surv(l, u, type = "interval2") ~ grp, data = miceData)
#> mtry = 1  OOB Brier score = 0.06497173 
#> Searching left ...
#> Searching right ...
```


NEWS.md
# ICcforest 0.5.0

* Added a `NEWS.md` file to track changes to the package.


R/ICcforest.R
#' Fit a conditional inference forest for interval-censored survival data
#'
#' An implementation of the random forest and bagging ensemble algorithms utilizing conditional 
#' inference trees as base learners for interval-censored survival data.
#' 
#' \code{ICcforest} returns an \code{ICcforest} object.
#' The object belongs to the class \code{ICcforest}, as a subclass of \code{\link[partykit]{cforest}}.
#' This function extends the conditional inference survival forest algorithm in 
#' \code{\link[partykit]{cforest}} to fit interval-censored survival data. 
#' 
#' 
#' @param formula a formula object, with the response being a \code{\link[survival]{Surv}} object, 
#' with form
#' 
#' 
#' \code{Surv(time1, time2, type="interval2")}.
#' @param data a data frame containing the variables named in \code{formula}.
#' @param na.action a function which indicates what should happen when the data contain 
#' missing values.
#' @param mtry number of input variables randomly sampled as candidates at each node for 
#' random forest like algorithms. The default \code{mtry} is tuned by \code{\link{tuneICRF}}.
#' @param ntree an integer, the number of the trees to grow for the forest. \code{ntree = 100L} is set 
#' by default.
#' @param perturb a list with arguments \code{replace} and \code{fraction} determining which 
#' type of resampling, with \code{replace = TRUE} referring to the n-out-of-n 
#' bootstrap and \code{replace = FALSE} referring to sample splitting. \code{fraction} is 
#' the proportion of observations to draw without replacement.
#' @param applyfun an optional \code{lapply}-style function with arguments 
#' \code{function(X, FUN, ...)}. 
#' It is used for computing the variable selection criterion. The default is to use the
#' basic \code{lapply} function unless the \code{cores} argument is specified (see below). 
#' See \code{\link[partykit]{ctree_control}}.
#' @param cores numeric. If set to an integer the \code{applyfun} is set to \code{\link[parallel]{mclapply}} 
#' with the desired number of cores. See \code{\link[partykit]{ctree_control}}.
#' @param trace whether to print the progress of the search of the optimal value of \code{mtry}
#' when \code{mtry} is not specified (see \code{\link{tuneICRF}}). \code{trace = TRUE} is set by default.
#' @param control a list of control parameters, see \code{\link[partykit]{ctree_control}}.
#' \code{control} parameters \code{minsplit}, \code{minbucket} have been adjusted from the 
#' \code{\link[partykit]{cforest}} defaults. Other default values correspond to those of the 
#' default values used by \code{\link[partykit]{ctree_control}}. 
#' @param suppress a logical specifying whether the messages from \code{\link[icenReg]{getFitEsts}} 
#' are suppressed. If \code{FALSE}, the messages are printed. \code{suppress = TRUE} is set by default.
#' @param ... additional arguments.
#' @keywords Ensemble method, conditional inference forest, interval-censored data
#' @return An object of class \code{ICcforest}, as a subclass of \code{\link[partykit]{cforest}}.
#' @import partykit
#' @import survival 
#' @import stats
#' @import utils
#' @import icenReg
#' @seealso \code{\link{predict.ICcforest}} for prediction, \code{\link{gettree.ICcforest}}
#' for individual tree extraction, and \code{\link{tuneICRF}} for \code{mtry} tuning. 
#' @examples
#' #### Example with miceData
#' library(icenReg)
#' data(miceData)
#' 
#' ## For ICcforest to run, Inf should be set to be a large number, for example, 9999999.
#' miceData$u[miceData$u == Inf] <- 9999999.
#' 
#' ## Fit an iterval-censored conditional inference forest
#' Cforest <- ICcforest(Surv(l, u, type = "interval2") ~ grp, data = miceData)
#' 
#' @export

ICcforest <- function(formula, data, mtry = NULL, ntree = 100L, applyfun = NULL, 
                      cores = NULL, na.action = na.pass, suppress = TRUE, trace = TRUE,
                      perturb = list(replace = FALSE, fraction = 0.632), 
                      control = partykit::ctree_control(teststat = "quad", testtype = "Univ", 
                                                        mincriterion = 0, saveinfo = FALSE,
                                                        minsplit = nrow(data) * 0.15, 
                                                        minbucket = nrow(data) * 0.06), ...){
  
  # package version dependency
  if (packageVersion("partykit") < "1.2.2") {
    stop("partykit >= 1.2.2 needed for this function.", call. = FALSE)
  }
  
  if (packageVersion("icenReg") < "2.0.8") {
    stop("icenReg >= 2.0.8 needed for this function.", call. = FALSE)
  }
  
  requireNamespace("inum")
  if (!requireNamespace("partykit", quietly = TRUE)) {
    stop("Package \"pkg\" needed for this function to work. Please install it.",
         call. = FALSE)
  }
  
  
  X <- data[,as.character(formula[[2]][[2]])]
  Y <- data[,as.character(formula[[2]][[3]])]
  
  if(sum(X==Y)){ ## add small noise to observed event case
    epsilon <- min(diff(sort(unique(c(X,Y)))))/20
    ID <- which(X==Y)
    data[ID,as.character(formula[[2]][[3]])] <- epsilon + data[ID,as.character(formula[[2]][[3]])]
  }

  ## x2 is Surv(Left,right,type="interval2") object
  .logrank_trafo <- function(x2){
    if(!(is.Surv(x2) && isTRUE(attr(x2, "type") == "interval"))){
      stop("Response must be a 'Survival' object with Surv(time1,time2,event) format")
    }
    
    # Fit IC survival curve
    Curve <- ic_np(x2[, 1:2])
    # get estimated survival
    Left <- 1 - getFitEsts(Curve, q = x2[, 1])
    Right <- 1 - getFitEsts(Curve, q = x2[, 2])
    
    Log_Left <- ifelse(Left<=0,0,Left*log(Left))
    Log_Right<- ifelse(Right<=0,0,Right*log(Right))
    result <- (Log_Left-Log_Right)/(Left-Right)
    
    return(matrix(as.double(result),ncol=1))
  }
  
  h2 <-function(y, x, start = NULL, weights, offset, estfun = TRUE, object = FALSE, ...) {
    if (all(is.na(weights))==1) weights <- rep(1, NROW(y))
    s <- .logrank_trafo(y[weights > 0,,drop = FALSE])
    r <- rep(0, length(weights))
    r[weights > 0] <- s
    list(estfun = matrix(as.double(r), ncol = 1), converged = TRUE)
  }
  
  if (is.null(mtry)) 
    mtry <- tuneICRF(formula, data, control = control, suppress = suppress, trace = trace)
  
  if (suppress == TRUE){
    invisible(capture.output(res <- partykit::cforest(formula, data, control = control, na.action = na.action, ytrafo = h2, 
                                                      mtry = mtry, ntree = ntree, applyfun = applyfun, cores = cores,
                                                      perturb = perturb, ...)))
  } else {
    res <- partykit::cforest(formula, data, control = control, na.action = na.action, ytrafo = h2, 
                             mtry = mtry, ntree = ntree, applyfun = applyfun, cores = cores,
                             perturb = perturb, ...)
  }
    
  
  # define a new class as a subclass of cforest
  class(res) <- c("ICcforest", class(res))
  return(res)
}


R/sbrier_IC.R
getFitEstsN <- function(fit, newdata = NULL, p = NULL, q = NULL){
  invisible(capture.output(res <- getFitEsts(fit, newdata, p, q)))
  return(res)
}

# get survival probability at time teval for ic_np, ic_sp, ic_par or ic_bayes object
getFitEsts_Surv <- function(Curve, newData = NULL, teval){
  if (is.null(newData)){
    Shat <- 1-getFitEstsN(fit = Curve, q = teval)
  } else {
    Shat <- 1-getFitEstsN(fit = Curve, newdata = newData, q = teval)
  }
  return(Shat)
}

#' Model Fit For Interval-Censored Data
#'
#' Compute the (integrated) Brier score to evaluate the model fit for interval-censored survival data.
#'
#' @param obj an object of class \code{\link[survival]{Surv}}.
#' @param pred predicted values. This can be a matrix of survival probabilities evaluated
#' at a sequence of time points for a set of new data, a list of \code{\link[survival]{survfit}}
#' objects, a list \code{\link[icenReg]{ic_np}}
#' objects, or a list of \code{\link[icenReg]{ic_sp}} objects.
#' @param btime a vector of length two indicating the range of times that the scores are computed on.
#' The default \code{btime} is set to be the vector of the smallest and the largest values among
#' all left and right endpoints given in \code{obj}.
#' @param type a character string denoting the type of scores returned. For \code{"IBS"},
#' the integrated Brier score over the \code{btime} is returned. For \code{"BS"}, the
#' Brier score at every left and right endpoint of all censoring intervals that lie within
#' \code{btime} is returned.
#' @keywords Brier score, integrated Brier score
#' @return
#' If \code{type = "IBS"}, this returns the integrated Brier score.
#' @return
#' If \code{type = "BS"}, this returns the Brier scores.
#' @import partykit
#' @import icenReg
#' @importFrom survival Surv
#' @import stats
#' @import ipred
#' @import utils
#' @references S. Tsouprou. Measures of discrimination and predictive accuracy for interval-censored
#' data. Master thesis, Leiden University. https://www.math.leidenuniv.nl/scripties/MasterTsouprou.pdf.
#' @examples
#' ### Example with dataset miceData
#' library(survival)
#' library(icenReg)
#' data(miceData)
#'
#' ## For proper evaluation, Inf should be set to be a large number, for example, 9999999.
#' idx_inf <- (miceData$u == Inf)
#' miceData$u[idx_inf] <- 9999999.
#'
#' obj <- Surv(miceData$l, miceData$u, type = "interval2")
#'
#' ## Model fit for an NPMLE survival curve with survfit
#' pred <- survival::survfit(formula = Surv(l, u, type = "interval2") ~ 1, data = miceData)
#' # Integrated Brier score up to time = 642
#' sbrier_IC(obj, pred, btime = c(0, 642), type = "IBS")
#'
#' ## Model fit for a semi-parametric model with icenReg::ic_sp()
#' pred <- icenReg::ic_sp(formula = Surv(l, u, type = "interval2") ~ 1, data = miceData)
#' # Integrated Brier score up to the largest endpoints of all censoring intervals in the dataset
#' sbrier_IC(obj, pred, type = "IBS")
#'
#' ## Model fit for an NPMLE survival curve with icenReg::ic_np()
#' pred <- icenReg::ic_np(miceData[,c('l', 'u')])
#' # Brier score computed at every left and right endpoints of all censoring intervals in the dataset
#' sbrier_IC(obj, pred, type = "BS")
#'
#'
#'
#' @export

sbrier_IC <- function(obj, pred, btime = range(as.numeric(obj[,1:2])), type = c("IBS","BS")) {
  if (missing(type))
    stop("The value of 'type' is not specified.")

  if (!inherits(obj, "Surv"))
    stop("obj is not of class Surv")
  class(obj) <- NULL
  # number of obs
  N <- nrow(obj)

  intL <- obj[, 1]
  intR <- obj[, 2]
  # the observation time in test set
  time <- c(intL,intR)

  if (is.null(btime))
    stop("btime not given")
  if (length(btime) < 1)
    stop("btime not given")
  if (length(btime) != 2) {
    stop("btime should be a vector of two indicating the range of the times")
  } else {
    if (btime[1] < min(time))
      warning("btime[1] is smaller than min(time)")
    if (btime[2] > max(time))
      warning("btime[2] is larger than max(time)")
    # all the observation times in the range of obs times
    btime <- unique(time[time >= btime[1] & time <= btime[2]])
    btime <- sort(btime)
  }

  ptype <- class(pred)[[1]]
  if (is.null(ptype)) {
    if (is.vector(pred))
      ptype <- "vector"
    if (is.list(pred))
      ptype <- "list"
  }

  if (ptype == "numeric" && is.vector(pred))
    ptype <- "vector"

  survs <- NULL
  switch(ptype, survfit = {
    survs <- try(ipred::getsurv(pred, time), silent=T)
    if(class(survs) == "try-error"){
      stop("please check whether the package 'ipred' is installed")
    }
    survs <- matrix(rep(survs, N), nrow = length(time))
  }, ic_np = {

    survs <- try(getFitEsts_Surv(Curve = pred, teval = time), silent=T)
    if(class(survs) == "try-error"){
      stop("please check whether the package 'icenReg' is installed")
    }
    survs <- matrix(rep(survs, N), nrow = length(time))
  }, ic_ph = {
    survs <- try(getFitEsts_Surv(Curve = pred, teval = time), silent=T)
    if(class(survs) == "try-error"){
      stop("please check whether the package 'icenReg' is installed")
    }
    survs <- matrix(rep(survs, N), nrow = length(time))
  }, list = {
    if (!inherits(pred[[1]], "survfit") && !inherits(pred[[1]], "ic_ph") && !inherits(pred[[1]], "ic_np"))
      stop("pred is not a list of survfit/ic_ph/ic_np objects; \n  if pred is an ICcforest prediction object, please make sure it is created using the setting type = 'prob'")
    if (length(pred) != N) stop("pred must be of length(time)")
    if (inherits(pred[[1]], "survfit")){
      M.list <- try(lapply(pred, ipred::getsurv, times = time),silent=T)
      if (class(M.list) == "try-error"){
        stop("please check whether the package 'ipred' is installed")
      }
      survs <- matrix(unlist(lapply(M.list, function(x) x)),
                      nrow = length(time), ncol = N) ## each obs in one column
    } else if (inherits(pred[[1]], "ic_np") || (inherits(pred[[1]], "ic_ph"))){
      M.list <- try(lapply(pred, getFitEsts_Surv, teval = time), silent=T)
      if(class(M.list) == "try-error"){
        stop("please check whether the package 'icenReg' is installed")
      }
      survs <- matrix(unlist(lapply(M.list, function(x) x)),
                      nrow = length(time), ncol = N) ## each obs in one column
    }
  }, vector = {
    if (length(pred) != N) stop("pred must be of length(time)")
    if (length(time) != 1) stop("cannot compute integrated Brier score with pred; \n  if pred is an ICcforest prediction object, please make sure it is created using the setting type = 'prob'")
    survs <- pred
  }, matrix = {
    if (all(dim(pred) == c(length(time), N))) survs <- pred
    else stop("wrong dimensions of pred; \n  if pred is an ICcforest prediction object, please make sure it is created using the setting type = 'prob'")
  })
  if (is.null(survs))
    stop("unknown type of pred")

  bsc <- rep(0, length(time))

  for (i in 1:length(time)) {
    bsc_temp <- matrix(0, nrow = N, ncol = 1)
    k = 0
    for (j in 1:N){
      St <- survs[i,j]
      if (time[i] <= intL[j]){
        IY = 1
      } else if(time[i] > intR[j]){
        IY = 0
      } else {
        Sr <- survs[N+j, j]
        Sl <- survs[j, j]
        if (Sl != Sr){
          IY = (St - Sr)/(Sl - Sr)
        } else {
          IY = NULL
        }
      }
      if (is.null(IY)==0){
        k = k + 1
        bsc_temp[k] = (IY - St)^2
      }
    }
    bsc[i] = mean(bsc_temp[1:k])
  }

  # get bsc and time ordered
  ot = order(time)
  bsc = bsc[ot]
  time = time[ot]

  # get bsc and time unique
  unik <- !duplicated(time)
  bsc <- bsc[unik]
  time <- time[unik]

  # bsc now corresponds with btime
  bsc <- bsc[time<=max(btime)]

  if (type == "IBS"){
    # apply trapezoid rule
    idx <- 2:length(btime)
    RET <- diff(btime) %*% ((bsc[idx - 1] + bsc[idx])/2)
    # the largest is much larger than the second largest
    if (max(btime)/tail(btime, n=2)[1] > 1000 && tail(bsc, n=1)==0 ){
      RET <- RET - (tail(btime, n=1) - tail(btime, n=2)[1])*((tail(bsc, n=1) + tail(bsc, n=2)[1])/2)
      btime <- head(btime, n = -1)
    }
    RET <- RET/diff(range(btime))
    names(RET) <- "integrated Brier score"
    attr(RET, "time") <- range(btime)
  } else if (type == "BS"){
    RET <- bsc
    names(RET) <- btime
    attr(RET, "type") <- "Brier score"
  } else {
    stop("unknown type of results")
  }
  RET
}


R/ICcforest-package.R
#' Construct a conditional inference forest model for interval-censored survival data
#' 
#' Construct a conditional inference forest model for interval-censored survival data.
#' The main function of this package is \code{\link{ICcforest}}.
#' 
#' \subsection{Problem setup and existing methods}{
#' In many situations, the survival time cannot be directly observed and it is only 
#' known to have occurred in an interval obtained from a sequence of examination times. 
#' Methods like the Cox proportional hazards model rely on restrictive assumptions such as 
#' proportional hazards and a log-linear relationship between the hazard function and 
#' covariates. Furthermore, because these methods are often parametric, nonlinear effects 
#' of variables must be modeled by transformations or expanding the design matrix to 
#' include specialized basis functions for more complex data structures in real world 
#' applications. The function \code{\link[LTRCtrees]{ICtree}} in the \code{LTRCtrees}
#' package provides a conditional inference tree method for interval-censored survival data, 
#' as an extension of the conditional inference tree method \code{\link[partykit]{ctree}}
#' for right-censored data. Tree estimators are nonparametric and as such often exhibit 
#' low bias and high variance. Ensemble methods like bagging and random forest can 
#' reduce variance while preserving low bias.}
#' 
#' \subsection{ICcforest model}{
#' This package implements \code{\link{ICcforest}}, which extends the conditional inference forest 
#' (see \code{\link[partykit]{cforest}}) to interval censored data. \code{\link{ICcforest}} uses 
#' conditional inference survival trees (see \code{\link[LTRCtrees]{ICtree}}) as base learners. 
#' The main function \code{\link{ICcforest}} fits a 
#' conditional inference forest for interval-censored survival data, with parameter 
#' \code{mtry} tuned by \code{\link{tuneICRF}}; \code{\link{gettree.ICcforest}} extracts 
#' the i-th individual tree from the established \code{ICcforest} objects; and 
#' \code{\link{predict.ICcforest}} computes predictions from \code{ICcforest} objects.}
#' 
#' @seealso \code{\link{ICcforest}, \link{gettree.ICcforest}, \link{predict.ICcforest},  
#' \link{tuneICRF}, \link{sbrier_IC}}
#' @docType package
#' @keywords package
#' @name ICcforest-package
NULL

R/gettree.ICcforest.R
#' Extract an individual tree from an ICcforest object
#'
#' Extract the i-th individual tree from the established ICcforest. The resulting object can be
#' printed or plotted, and predictions can be made using it.
#' 
#' @param object an object as returned by \code{\link{ICcforest}}.
#' @param tree an integer, the number of the tree to extract from the forest.
#' @param ... additional arguments.
#' @return An object of class \code{\link[partykit]{party}}.
#' @import partykit
#' @import icenReg
#' @importFrom survival Surv
#' @import stats
#' @import utils
#' @examples
#' 
#' #### Example with dataset miceData
#' library(icenReg)
#' data(miceData)
#' 
#' ## For ICcforest to run, Inf should be set to be a large number, for example, 9999999.
#' idx_inf <- (miceData$u == Inf)
#' miceData$u[idx_inf] <- 9999999.
#' 
#' ## First, fit an iterval-censored conditional inference forest
#' Cforest <- ICcforest(formula = Surv(l,u,type="interval2")~grp, data = miceData, ntree = 50L)
#' ## Extract the 50-th tree from the forest
#' plot(gettree(Cforest, tree = 50L))
#' 
#' @export
#'
#'

gettree.ICcforest <- function(object, tree = 1L, ...) {
  ft <- object$fitted
  ft[["(weights)"]] <- object$weights[[tree]]
  ret <- partykit::party(object$nodes[[tree]], data = object$data, fitted = ft)
  ret$terms <- object$terms
  class(ret) <- c("constparty", class(ret))
  return(ret)
}



R/predict.ICcforest.R
#' Predict from an ICcforest model
#' 
#' Compute predictions from ICcforest objects.
#' 
#' @param object an object as returned by \code{\link{ICcforest}}.
#' @param newdata an optional data frame containing test data.
#' @param type a character string denoting the type of predicted value returned. 
#' 
#' 
#' For \code{"type = response"}, the mean of a numeric response, the median survival time for 
#' the interval-censored response is returned. For \code{"type = prob"}, a list with the survival 
#' function constructed using the non-parametric maximum likelihood estimator for each observation 
#' is returned. \code{"type = weights"} returns an integer vector of prediction weights. 
#' For \code{type = "node"}, a list of terminal node ids for each of the trees in the 
#' forest is returned.
#' @param FUN a function to compute summary statistics. Predictions for each node must be 
#' computed based on arguments \code{(y, w)} where \code{y} is the response and \code{w} are case 
#' weights. 
#' @param OOB a logical specifying whether out-of-bag predictions are desired
#' 
#' 
#' (only if \code{newdata = NULL}).
#' @param simplify a logical indicating whether the resulting list of predictions should be 
#' converted to a suitable vector or matrix (if possible), see \code{\link[partykit]{cforest}}.
#' @param scale a logical indicating scaling of the nearest neighbor weights by the sum of weights 
#' in the corresponding terminal node of each tree, see \code{\link[partykit]{cforest}}.
#' @param suppress a logical specifying whether the messages from \code{\link[icenReg]{getFitEsts}} 
#' are suppressed. If \code{FALSE}, the messages are printed. \code{suppress = TRUE} is set by default.
#' @param ... additional arguments.
#' @return An object of class \code{ICcforest}, as a subclass of \code{\link[partykit]{cforest}}.
#' @import partykit
#' @import survival 
#' @import stats
#' @import utils
#' @import icenReg
#' @seealso \code{\link{sbrier_IC}} for evaluation of model fit for interval-censored data
#' @examples 
#' library(icenReg)
#' data(miceData)
#' 
#' ## For ICcforest to run, Inf should be set to be a large number, for example, 9999999.
#' miceData$u[miceData$u == Inf] <- 9999999.
#' 
#' ## First, fit an iterval-censored conditional inference forest
#' Cforest <- ICcforest(formula = Surv(l,u,type="interval2")~grp, data = miceData)
#' ## Predict the survival function constructed using the non-parametric maximum likelihood estimator
#' Pred <- predict(Cforest, type = "prob")
#' 
#' ## Out-of-bag prediction of the median survival time
#' PredOOB <- predict(Cforest, type = "response", OOB = TRUE)
#' 
#' @export
#' 
#' 

predict.ICcforest <- function(object, newdata = NULL, OOB = FALSE, suppress = TRUE,
                              type = c("response", "prob", "weights", "node"),
                              FUN = NULL, simplify = TRUE, scale = TRUE, ...){
  
  # package version dependency
  if (packageVersion("partykit") < "1.2.2") {
    stop("partykit >= 1.2.2 needed for this function.", call. = FALSE)
  }
  
  if (packageVersion("icenReg") < "2.0.8") {
    stop("icenReg >= 2.0.8 needed for this function.", call. = FALSE)
  }
  
  if (missing(type)){
    stop("The value of 'type' is not specified.")
  }
  
  pred_Surv_prob <- function(y, w) {
    if (length(y) == 0) return(NA)
    idx = which(w>0)
    y = y[idx]
    w = w[idx]
    yy = as.matrix(y[,1:2])
    ic_np(yy, weights = w)
  }
  
  pred_Surv_response <- function(y, w) {
    if (length(y) == 0) return(NA)
    getFitEsts(pred_Surv_prob(y, w))
  }
  
  if (is.null(FUN)) 
    Fun <- if (type == "response") pred_Surv_response else pred_Surv_prob
  else
    Fun <- NULL
  
  if (suppress == TRUE){
    invisible(capture.output(res <- partykit::predict.cforest(object = object, newdata = newdata, type = type, 
                                                              OOB = OOB, simplify = simplify, scale = scale,
                                                              FUN = Fun, ...)))
  } else{
    res <- partykit::predict.cforest(object = object, newdata = newdata, type = type, 
                                     OOB = OOB, simplify = simplify, scale = scale,
                                     FUN = Fun, ...)
  }
  return(res)
}




R/tuneICRF.R
#' Tune mtry to the optimal value with respect to out-of-bag error for an ICcforest model
#' 
#' Starting with the default value of mtry, search for the optimal value 
#' (with respect to Out-of-Bag error estimate) of mtry for ICcforest.
#' 
#' @param formula a formula object, with the response being a \code{\link[survival]{Surv}} 
#' object, with form 
#' 
#' 
#' \code{Surv(time1, time2, type="interval2")}.
#' @param data a data frame containing the variables named in \code{Formula}.
#' @param mtryStart starting value of \code{mtry}; default is \code{sqrt(nvar)}.
#' @param stepFactor at each iteration, \code{mtry} is inflated (or deflated) by this value.
#' @param control a list with control parameters, see \code{\link[partykit]{cforest}}. 
#' The default values correspond to those of the default values used by \code{\link{ICcforest}}.
#' @param ntreeTry number of trees used at the tuning step.
#' @param trace whether to print the progress of the search. \code{trace = TRUE} is set by default.
#' @param plot whether to plot the out-of-bag error as a function of \code{mtry}.
#' @param doBest whether to run an ICcforest using the optimal mtry found.
#' @param suppress a logical specifying whether the messages from \code{\link[icenReg]{getFitEsts}} 
#' are suppressed. If \code{FALSE}, the messages are printed. \code{suppress = TRUE} is set by default.
#' @param ... additional arguments.
#' @keywords mtry, out-of-bag errors, brier score
#' @return 
#' If \code{doBest=FALSE} (default), this returns the optimal mtry value of those searched.
#' @return 
#' If \code{doBest=TRUE}, this returns the ICcforest object produced with the optimal mtry.
#' @import partykit
#' @importFrom survival Surv
#' @importFrom graphics axis
#' @import stats
#' @import utils
#' @import icenReg
#' @seealso \code{\link{sbrier_IC}} for evaluation of model fit for interval-censored data 
#' when searching for the optimal value of \code{mtry}.
#' @examples
#' ### Example with dataset tandmob2
#' library(icenReg)
#' data(miceData)
#' 
#' ## For ICcforest to run, Inf should be set to be a large number, for example, 9999999.
#' miceData$u[miceData$u == Inf] <- 9999999.
#'
#' ## Create a new variable to be selected from
#' miceData$new = rep(1:4)
#' 
#' ## Tune mtry 
#' mtryTune <- tuneICRF(Surv(l, u, type = "interval2") ~ grp + new, data = miceData)
#' 
#' @export

tuneICRF <- function(formula, data, mtryStart = NULL, stepFactor = 1.5, ntreeTry = 100L, 
                     control = partykit::ctree_control(teststat = "quad", testtype = "Univ", 
                                                       mincriterion = 0, saveinfo = FALSE,
                                                       minsplit = nrow(data) * 0.15, 
                                                       minbucket = nrow(data) * 0.06), 
                     suppress = TRUE, trace = TRUE, plot = FALSE, doBest = FALSE) {
  # package version dependency
  if (packageVersion("partykit") < "1.2.2") {
    stop("partykit >= 1.2.2 needed for this function.", call. = FALSE)
  }
  
  if (packageVersion("icenReg") < "2.0.8") {
    stop("icenReg >= 2.0.8 needed for this function.", call. = FALSE)
  }
  
  # number of the variables
  allX <- substring(formula,1)[[3]]
  nameX <- strsplit(gsub("\\+", '', allX)," ")[[1]]
  nameX <- nameX[nameX!=""]
  nvar = length(nchar(nameX))
  if (is.null(mtryStart)){
    mtryStart = ceiling(sqrt(nvar))
  } 
  
  # integrated Brier score of out-of-bag samples for a mtry value at test 
  errorOOB_mtry <- function(formula, data, mtryTest, ntreeTry, control, suppress){
    cfOOB <- ICcforest(formula, data, mtry = mtryTest, ntree = ntreeTry, control = control, suppress = suppress)
    predOOB <- predict(object = cfOOB, newdata = NULL, OOB = TRUE, type = "prob", suppress = suppress)
    aSurv <- data[,as.character(formula[[2]][[2]])]
    bSurv <- data[,as.character(formula[[2]][[3]])]
    
    testObj <- survival::Surv(aSurv,bSurv,type = "interval2")
    errorOOB <- unname(sbrier_IC(testObj, predOOB, type = "IBS")[1])
    
    rm(cfOOB)
    rm(predOOB)
    return(errorOOB)
  }
  
  
  # errorOld
  errorOld <- errorOOB_mtry(formula, data, mtryTest=mtryStart, ntreeTry, control, suppress = suppress)

  if (errorOld < 0) stop("Initial setting gave 0 error and no room for improvement.")
  if (trace) {
    cat("mtry =", mtryStart, " OOB Brier score =",
        errorOld, "\n")
  }
  
  oobError <- list()
  oobError[[1]] <- errorOld
  names(oobError)[1] <- mtryStart  
  
  for (direction in c("left", "right")) {
    if (trace) cat("Searching", direction, "...\n")
    mtryCur <- mtryStart
    while (mtryCur != nvar) {
      mtryOld <- mtryCur
      mtryCur <- if (direction == "left") {
        max(1, ceiling(mtryCur / stepFactor))
      } else {
        min(nvar, floor(mtryCur * stepFactor))
      }
      if (mtryCur == mtryOld) break
      
      errorCur <- errorOOB_mtry(formula, data, mtryTest = mtryCur, ntreeTry, control, suppress = suppress)

      if (trace) {
        cat("mtry =",mtryCur, "\tOOB error =",errorCur, "\n")
      }
      oobError[[as.character(mtryCur)]] <- errorCur
      errorOld <- errorCur
    }
  }
  mtry <- sort(as.numeric(names(oobError)))
  res_all <- unlist(oobError[as.character(mtry)])
  res_all <- cbind(mtry = mtry, OOBError = res_all)
  res <- res_all[which.min(res_all[,2]), 1]
  
  if (plot) {
    res = res_all
    plot(res_all, xlab = expression(m[try]), ylab = "OOB Error", type = "o", log = "x", xaxt = "n")
    axis(1, at=res_all[,"mtry"])
  }
  
  if (doBest) 
    res <- ICcforest(formula, data, mtry = res, ntree = ntreeTry, control = control, suppress = suppress)
  
  return(res)
}
